### ISSUE TITLE
git archive fails on LFS when a subdirectory is specified

### ISSUE THREAD

**Issue by mcclure** [git-lfs/git-lfs#3769]
In issue #1322 I was told I should expect git archive to work with LFS as long as I have a working copy checked out locally. In my testing (git 2.22.0 on Ubuntu on Windows Subsystem for Linux, I originally saw this with git-lfs 2.3.4 but I upgraded to git-lfs 2.8.0 for this test) I see that this is true if I `git archive HEAD` but not true if I `git archive HEAD:some/subdirectory`.

#### Repro steps

Clone https://github.com/mcclure/git-lfs-demo . This project contains two text files stored in lfs and a directory. From inside the directory I run:

```
mkdir _archive_no_dir
git archive HEAD | tar -x -C _archive_no_dir
cat _archive_no_dir/dir/testfile2.txt
mkdir _archive_with_dir
git archive HEAD:dir | tar -x -C _archive_with_dir
cat _archive_with_dir/testfile2.txt
```

Expected behavior is that both testfile2s would be the same. Observed behavior is that the first testfile2 is correct and the second testfile2 is a git-lfs stub file.

#### Context

The reason I encountered this is I have a project with many art assets in a complex directory structure. My build script has a step where it archives just the assets directory, so I can be sure the build product contains only files that are in source control. I have recently switched to git-lfs and now my builds are broken.

Is this supposed to work? Is there a workaround?

**Comment by bk2204** [comment]
Hey,

I've downloaded the test repo and I can confirm the behavior you're seeing.

The way Git produces archives with `git archives` depends on the `.gitattributes` file in the tree (directory) it's archiving. When you ask it to archive `HEAD`, the tree has a `.gitattributes` file at its root, so it honors it and calls Git LFS to translate your files for you. When you archive `HEAD:dir`, there is no `.gitattributes` file, so Git doesn't call Git LFS, and you end up with a pointer file, as you noticed.

I think the Git folks would probably say that this is working the way that's intended, although it produces some unfortunate results, not only for Git LFS, but also for things like line endings and encodings, which are also controlled by the `.gitattributes` file. There is, however, a workaround which may be useful for your use case.

It's possible to place a `.gitattributes` file in any directory of the tree, and such a file affects that directory and any subdirectories (but not any parents). So if you have all of your art assets in a directory (let's call it `assets`) and none of them in any other directory, then you can simply move the relevant `.gitattributes` entries to `assets/.gitattributes`. When you run `git archive HEAD:assets`, that file will be honored, and your assets will end up as their original selves (non-pointer files).

If you have patterns that apply to files both in and out of your assets directory, then it may be necessary to duplicate some of these patterns in your assets `.gitattributes` if you want to archive them in this way.

**Comment by miccoli** [comment]
A possible quick workaround is
```
mkdir _archive_with_dir_workaround
git archive HEAD dir | tar -x --strip-components=1 -C _archive_with_dir_workaround
cat _archive_with_dir_workaround/testfile2.txt 
```

In other words
-  `git archive HEAD:dir | tar -x` 
-  `git archive HEAD dir | tar -x --strip-components=1`
produce the same same tree, but the latter commands *sees* the toplevel `.gitattributes` file in `HEAD` and applies the lfs filter (although the `.gitattribute` file itself is not included in the archive)

Other options are using the `git archive --worktree-attributes` which honours the *checked out* `.gitattributes` (not the one in HEAD) or creating a `$GIT_DIR/info/attributes` file. These solutions seem to me less obvious than the first one.

BTW, this behaviour was completely unexpected to me, and I would consider it a core git bug.

**Comment by mcclure** [comment]
> BTW, this behaviour was completely unexpected to me, and I would consider it a core git bug.

I would agree, it may be that "git is working properly" or there are reasons why this is an inevitable interop problem between git and git-lfs, but from an end user perspective there is no real reason to think of git and git-lfs as two different projects…

The very simple workaround I wound up using in my personal project was `git archive HEAD | tar -x -C $DESTDIR/Data subdirectory/with/assets`. So in this case git archives the entire repo but tar ignores everything but subdirectory/with/assets. This does mean the files in the end go to `$DESTDIR/Data/subdirectory/with/assets` but I can just write my scripts to anticipate that.
